library/core/src/sync/atomic.rs

```diff
@@ -92,13 +92,14 @@
 //! "sufficiently small" depends on the architecture:
 //!
 //! | Target architecture | Maximal atomic `load` size that is guaranteed read-only |
-//! |-----------|---------|
-//! | `x86`     | 4 bytes |
-//! | `x86_64`  | 8 bytes |
-//! | `arm`     | 4 bytes |
-//! | `aarch64` | 8 bytes |
-//! | `riscv32` | 4 bytes |
-//! | `riscv64` | 8 bytes |
+//! |---------------|---------|
+//! | `x86`         | 4 bytes |
+//! | `x86_64`      | 8 bytes |
+//! | `arm`         | 4 bytes |
+//! | `aarch64`     | 8 bytes |
+//! | `riscv32`     | 4 bytes |
+//! | `riscv64`     | 8 bytes |
+//! | `powerpc64le` | 8 bytes |
 //!
 //! Any atomic `load` on read-only memory larger than the given size are Undefined Behavior. For
```

Comment:
I would argue that, rather than always being undefined behavior, they are instead sometimes undefined behavior depending on enabled/default target features and OS libraries. e.g. on `powerpc64le-unknown-linux-gnu` and/or when the `quadword-atomics` feature is enabled, 16-byte atomic loads are perfectly fine.

Reply:
Saying "it's not UB sometimes" is useless, we'd have to list the concrete guarantees we are making. I don't feel like it's worth the effort of bloating this table to a huge monster that covers every single combination of architecture, OS, and set of enabled target features...

Note that this table doesn't say 16-byte atomic loads are forbidden. It says 16-byte atomic loads *from read-only memory* are forbidden.

Reply:
> Saying "it's not UB sometimes" is useless, we'd have to list the concrete guarantees we are making.

the concrete guarantees are not explicitly specified, but mostly follow target features.

I think declaring everything not explicitly listed here as UB is taking it too far, because that means the following is UB even though most any reasonable person would expect it to be fine:
```rust
#[cfg(target_arch = "powerpc64le")]
#[cfg(target_feature = "quadword-atomics")]
pub fn f(v: &u128) -> u128 {
    unsafe { (*(v as *const _ as *const AtomicU128)).load(Ordering::Relaxed) }
}
```

Reply:
I agree with Jacob that we shouldn't say this is UB if it is okay on some targets. It sounds like the situation is more like some combination of "target-dependent" and "unspecified", with some targets getting clearer upper and lower bounds. I think a better way to phrase it is that it is UB to use an atomic if the load lowers to a read-write atomic operation on read-only memory, and whether this actually happens is target-dependent with some specific cases being specified explicitly.

Reply:
I'm strongly opposed to saying things like "UB unless the backend does X". We don't make any guarantees about what the backend does unless documented otherwise, and this PR *is* that documentation. So such a statement doesn't actually make any more code well-defined, while at the same time encouraging people to look at what the backend does and depend on it as if it were a stable guarantee.

Reply:
We could say "for architectures not in the table, the maximum size for which atomic loads are read-only is not specified can cannot be relied upon". That's logically equivalent to saying that the threshold size is 0, which is logically equivalent to what the PR already says.

Reply:
> I'm strongly opposed to saying things like "UB unless the backend does X".

I think it should be read more as "UB unless predicate P holds", where P is only loosely constrained and is otherwise target-dependent. It's not actually compiler-dependent (or at least it shouldn't be). The bounds on P are there to help people reason about the common targets, and everything else seems like a spec hole to me, although we may for pragmatic reasons hold off on defining it.

> IOW, to make your statement useful, we need a table of how atomic loads get lowered on each target, and we need to say that if for a given target there's no entry in the table then one may not make any assumptions about how it is lowered.

Yes, that is the intent. If there are no bounds specified, then the programmer is not allowed to assume it is not UB and hence can't rely on it. This is not the same as it actually being UB though, just like relying on unspecified layout details is not UB if you are able to guess correctly by "insider information". In this case, this would cover things like Jacob's `quadword-atomics` feature, until/unless rustc specifies this feature more precisely or adds it to the table.

Reply:
I guess we can use wording similar to what I used in the ABI-guarantees PR.
```
//! Atomics loads that are larger than this threshold (and *all* atomic loads on targets not listed
//! in the table) might still be read-only under certain conditions, but that is not a stable
//! guarantee and should not be relied upon.
```